sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/MethodOverloadsShouldBeGroupedBase.cs

```diff
@@ -24,99 +24,111 @@
                     {
-                        if (!key.Equals(previousKey, CaseSensitivity))
+                        if (!current.Equals(previous))
                         {
-                            currentList.Add(member);
+                            values.Add(current);
                         }
                     }
                     else
                     {
-                        misplacedOverloads.Add(key, new List<TMemberDeclarationSyntax> { member });
+                        misplacedOverloads.Add(current, new List<MemberInfo> { current });
                     }
-                    previousKey = key;
+                    previous = current;
                 }
                 else
                 {
-                    previousKey = null;
+                    previous = null;
                 }
             }
-            return misplacedOverloads;
+            return misplacedOverloads.Values.Where(x => x.Count > 1).ToArray();
         }
 
-        private string GetMethodName(TMemberDeclarationSyntax member, bool caseSensitive)
+        protected class MemberInfo
         {
-            var nameSyntaxNode = GetNameSyntaxNode(member);
-            if (caseSensitive)
+
+            public readonly TMemberDeclarationSyntax Member;
+            public readonly string Accessibility;
+            public readonly SyntaxToken NameSyntax;
+            public readonly bool IsStatic;
+            public readonly bool IsAbstract;
+
+            private readonly bool isCaseSensitive;
+
+            public MemberInfo(SyntaxNodeAnalysisContext context, TMemberDeclarationSyntax member, SyntaxToken nameSyntax, bool isStatic, bool isAbstract, bool isCaseSensitive)
+            {
+                Member = member;
+                Accessibility = context.SemanticModel.GetDeclaredSymbol(member)?.DeclaredAccessibility.ToString();
+                NameSyntax = nameSyntax;
+                IsStatic = isStatic;
+                IsAbstract = isAbstract;
+                this.isCaseSensitive = isCaseSensitive;                
+            }
+            
+            public override bool Equals(object obj)
             {
-                return nameSyntaxNode?.ValueText;
+                // Groups that should be together are defined by accessibility, abstract, static and member name #4136
+                return obj is MemberInfo info
+                    && Accessibility == info.Accessibility
+                    && NameSyntax.ValueText.Equals(info.NameSyntax.ValueText, isCaseSensitive && info.isCaseSensitive ? StringComparison.InvariantCulture : StringComparison.InvariantCultureIgnoreCase)
+                    && IsStatic == info.IsStatic
+                    && IsAbstract == info.IsAbstract;
             }
-            else
+
+            public override int GetHashCode()
             {
-                return nameSyntaxNode?.ValueText?.ToLowerInvariant();
+                return NameSyntax.ValueText.ToLowerInvariant().GetHashCode();
```

Comment:
It seems strange to have the hashcode being generated from a single field, when the `Equals` method is using 4 of them

Reply:
I wanted them to end up in same bucket and let Equals to deal with the rest. I didn't want to duplicate the full Equals/Grouping logic here with that ugly and long concatenation for generating the key.